### PR TITLE
Implicitly add PackageReference for NETStandard.Library if targeting .NET Standard

### PR DESCRIPTION
**Pull request by dsplaisted**
This PR will implicitly add a `PackageReference` to NETStandard.Library if the project is targeting .NETStandard and there isn't already a `PackageReference` to NETStandard.Library.  This makes project files more succinct and avoids confusion about why there are different versions of .NET Standard in a project file: for example "netstandard1.4" in the `TargetFramework` property but version "1.6.0" in the `PackageReference`.

### CODE REVIEW

**Comment by dsplaisted** [comment]
@davidfowl @ericstj @terrajobst In addition to making the NETStandard.Library `PackageReference` implicit, I'd like to do the same thing with Microsoft.NETCore.App.  Does that package work the same way as NETStandard.Library, where you can always reference the latest version of the package but target a lower version of the framework?  Can I have a `PackageReference` to Microsoft.NETCore.App version 1.1.0, but have my `TargetFramework` be netcoreapp1.0?

@davidfowl also [brought up](https://github.com/dotnet/sdk/issues/413#issuecomment-262719083) the issue of how this would work with patch versions of .NET Core.  If for Microsoft.NETCore.App, the package version and target framework version need to be the same, can I set the `TargetFramework` to netcoreapp1.0.1?  Or does the package with that version only support 1.0 as the target framework version?

**Comment by dsplaisted** [comment]
@rainersigwald @AndyGerlicher @cdmihai Is there a better way of checking whether a specific item exists than using two excludes [like I'm doing here](https://github.com/dsplaisted/sdk/blob/2d65e9dcbf5938b9ff82dd9ebf23519e6dfd9884/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets#L95-L103)?

**Comment by rainersigwald** [comment]
@dsplaisted Yes, using the `WithMetadataValue` [Item Function](https://docs.microsoft.com/en-us/visualstudio/msbuild/item-functions):

```xml
<Project>
  <ItemGroup>
    <I Include="foo" />
    <I Include="bar" />
    <T Include="@(I->WithMetadataValue('Identity', 'foo'))" />
  </ItemGroup>
  <Target Name="Build">
    <Message Importance="high"
             Text="I: @(I)" />
    <Message Importance="high"
             Text="T: @(T)" />
  </Target>
</Project>
```

```
s:\msbuild>msbuild test.proj
Microsoft (R) Build Engine version 15.1.371.0
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 12/2/2016 5:19:36 PM.
Project "s:\msbuild\test.proj" on node 1 (default targets).
Build:
  I: foo;bar
  T: foo
Done Building Project "s:\msbuild\test.proj" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.09
```

**Comment by davidfowl** [comment]
This needs to be coordinated with the nuget feature that allows specifying metadata on package references so the ui doesn't show updates for them. You don't want this abstraction breaking down when any action you take undoes the implicitness.

**Comment by davidfowl** [comment]
> @davidfowl @ericstj @terrajobst In addition to making the NETStandard.Library PackageReference implicit, I'd like to do the same thing with Microsoft.NETCore.App. Does that package work the same way as NETStandard.Library, where you can always reference the latest version of the package but target a lower version of the framework? Can I have a PackageReference to Microsoft.NETCore.App version 1.1.0, but have my TargetFramework be netcoreapp1.0?

Microsoft.NETCore.App should be exactly the same idea. Newer versions should work with older frameworks.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (davidfowl) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Hard coding 1.6.0? Don't we want to match the major and minor of the netstandard TFM if it's > 1.6.0?

Comment (DamianEdwards) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Not to mention [we're already at 1.6.1](http://www.nuget.org/packages/NETStandard.Library/1.6.1)

**Comment by emgarten** [comment]
On the nuget side https://github.com/NuGet/Home/issues/4044 is tracking the work for ignoring updates to references such as this.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (ericstj) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: I thought @terrajobst had a spec on this.

**Comment by dsplaisted** [comment]
@davidfowl 

> This needs to be coordinated with the nuget feature that allows specifying metadata on package references so the ui doesn't show updates for them. You don't want this abstraction breaking down when any action you take undoes the implicitness.

I've moved the implicit `PackageReferences` to the `.props` files, which means the behavior is now this:

- You get an implicit `PackageReference` for NETStandard.Library or Microsoft.NETCore.App if you are targeting .NETStandard or .NETCoreApp
- The version of the package reference is the greater of your target framework version and the highest version of that package that the Sdk knows about
- You can put a `PackageReference` to a different version of one of these packages in your project file, and it will override the implicit one from the `.props`
- The package (and available updates) will show up in the NuGet package manager UI.  If you use the package manager UI to upgrade the package, NuGet will put the `PackageReference` in your `.csproj` for you

To me, this seems like ideal behavior.  I don't think we need to hide updates to implicit package references from the UI if the UI knows how to correctly apply them.

@srivatsn @nguerrera for review

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: How does this work? Isn't TargetFrameorkIdentifier only set after props?

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: Oh because items are evaluated on a later pass? Very confusing.

**Comment by gulbanana** [comment]
This doesn't seem like it would always be ideal behaviour. Imagine a world where NETStandard.Library 2.0 has been released, and two developers in a team are sharing code:
- Developer A updates their cli/sdk, and the new version knows about package 2.0
- Developer A is now implicitly using 2.0 and begins to rely on new APIs
- Developer B hasn't noticed an update notification or hasn't had time to upgrade, so they're still using the previous SDK version
- Developer A's changes don't compile on Developer B's machine even though their project files are identical!

**Comment by DamianEdwards** [comment]
@gulbanana it shouldn't be based on the SDK you are using, but the TFM you're targeting. They're independent.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (davidfowl) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Is this a version comparison?

**Comment by gulbanana** [comment]
@damianedwards it's the greater-of behaviour dplaisted describes that concerns me a little. if the tfm was left at netstandard1.6 you'd get library 1.6 or 2.0 based on where it was built.

**Comment by DamianEdwards** [comment]
Indeed but the packages are pivoted internally by TFM, so it shouldn't
actually affect anything.

On Dec 6, 2016 9:42 PM, "gulbanana" <notifications@github.com> wrote:

@damianedwards <https://github.com/damianedwards> it's the greater-of
behaviour dplaisted describes that concerns me a little. if the tfm was
left at netstandard1.6 you'd get library 1.6 or 2.0 based on where it was
built.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/dotnet/sdk/pull/450#issuecomment-265361179>, or mute
the thread

**Comment by DamianEdwards** [comment]
But to your point, the version implied should be the latest version the SDK knows about that matches the TFM being targeted.

**Comment by davidfowl** [comment]
Higher is always safe but I agree different behavior based on the SDK installed isn't ideal.

> The version of the package reference is the greater of your target framework version and the highest version of that package that the Sdk knows about

It might be better to just use the major and minor of the TFM so that things are consistent no matter what the SDK knows about.

**Comment by DamianEdwards** [comment]
Yeah that's what I'm proposing.

**Comment by DamianEdwards** [comment]
Although I think it's fine to have it float the patch level to highest the SDK is aware of, maybe? Still potential for edge case differences, but vastly reduced and removes the need to have a PackageReference to get latest patch levels by default.

**Comment by gulbanana** [comment]
If it's an equal-minor check that sounds like it won't cause any serious problems. You could still get slightly weird behaviour when, like, two different people are making release of a nuget package, and its dependencies keep switching between 1.6.1 and 1.6.2 - but this isn't likely to cause real issues.

**Comment by davidfowl** [comment]
>Although I think it's fine to have it float the patch level to highest the SDK is aware of, maybe? Still potential for edge case differences, but vastly reduced and removes the need to have a PackageReference to get latest patch levels by default.

Yes that would be ok, though it would be great if we avoided patches in general 😄 /cc @terrajobst .

**Comment by DamianEdwards** [comment]
Sure, just don't have bugs in the BCL, easy.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Yes.  When you do a greater/less than comparison with strings, MSBuild tries to parse them as Versions and compare them as such.

**Comment by dsplaisted** [comment]
If you're worried about newer versions of the implicitly referenced packages behaving differently, you should also be worried about differences in behavior between newer versions of the .NET Sdk.  So you should pin the Sdk version and then you'll always get the same version of the NuGet packages too.

Right now you can use globals.json to specify which version of the .NET CLI should be used, and we are in the early stages of designing something similar that would refer to the MSBuild Sdks referenced with the `Sdk` attribute on the `Project` element.

So I think the logic to use the latest version of the package that the Sdk knows about is a good choice.  Does that make sense?

**Comment by davidfowl** [comment]
> If you're worried about newer versions of the implicitly referenced packages behaving differently, you should also be worried about differences in behavior between newer versions of the .NET Sdk. So you should pin the Sdk version and then you'll always get the same version of the NuGet packages too.

Seems orthogonal. Explicitly updating the SDK is one thing but it shouldn't affect my compilation references if I changed nothing else. Less variance is better here and we should try our best to keep the behavior as similar as possible even when the SDK revs with respect to the runtime behavior.

**Comment by dsplaisted** [comment]
> Explicitly updating the SDK is one thing but it shouldn't affect my compilation references if I changed nothing else

Updating to a newer version of the NETStandard.Library or Microsoft.NETCore.App packages shouldn't change your compilation references either, unless we make a mistake or fix a bug.

That said, I don't think it's super critical either way, so I'd be happy to change it to just use the version from the target framework version if there's a chance we can make the PackageReferences implicit for RC2.  We can always change teh logic later.

**Comment by davidfowl** [comment]
> Updating to a newer version of the NETStandard.Library or Microsoft.NETCore.App packages shouldn't change your compilation references either, unless we make a mistake or fix a bug.

I agree but we messed this up before, and we can do it again (it's actually broken today in pre release versions of Microsoft.NETCore.App 1.2.0-*). 

>That said, I don't think it's super critical either way, so I'd be happy to change it to just use the version from the target framework version if there's a chance we can make the PackageReferences implicit for RC2. We can always change the logic later.

Yea, I think we should just change it for now. Users can always override it anyways. This is a good change to make now so we can dogfood the experience.

File: TestAssets/TestProjects/SimpleDependencies/SimpleDependencies.csproj
Comment (nguerrera) [inline TestAssets/TestProjects/SimpleDependencies/SimpleDependencies.csproj]: Did you leave any tests with explicit package references overriding the default?

**Comment by nguerrera** [comment]
>  if there's a chance we can make the PackageReferences implicit for RC2

I don't think this should go in to RC2 so late.

**Comment by nguerrera** [comment]
One thing that was mentioned but maybe not called out explicitly enough is that while a change in the package version should not affect runtime or even compilation behaviour, it will change the nupkg dependencies produced by pack. Are we OK with that varying based on the SDK used?

**Comment by nguerrera** [comment]
I would really like to get @terrajobst's approval here to make sure we're aligned on the direction.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (davidfowl) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: lol, that is so random

**Comment by davidfowl** [comment]
When is this going in @dsplaisted ?

**Comment by dsplaisted** [comment]
Feedback from @terrajobst was that the default package reference in the SDK should always be to the same version for a given SDK (and package name).  So I need to make that change.

Also, when I tested the behavior of using the UI to update the NuGet package, I think it was writing out `<PackageReference Include="..."` when it should probably adding `<PackageReference Update="..."`.  So I need to investigate what's going on there too.

**Comment by dsplaisted** [comment]
I've updated this PR based on @terrajobst's feedback, and verified that upgrading or downgrading the package reference via the package manager UI works correctly.

The only thing I'm worried about is that, as discussed in #501, when you're targeting netcoreapp, you can't target a lower version of the framework and reference a later version of the Microsoft.NETCore.App package.  If we can't fix that then we may have to re-think how we handle an implicit package reference when targeting .NET Core.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (srivatsn) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: No check for DisableImplicitFrameworkReferences?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (srivatsn) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: Should this also be ReadOnly=true?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets
Comment (srivatsn) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Where is this property defined and what's the relevance of adding this property to this change?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: No.  I haven't yet written up the thinking behind this change, but basically since right now the Microsoft.NETCore.App version needs to match the TargetFrameworkVersion, the idea is to hide the package reference entirely.  The NETStandard.Library package supports lower versions of the target framework, so it's OK to surface updates to that package in the UI.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: No (and that's called out in the comment), because this package reference should be hidden from the developer as much as possible.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: There's a bit of explanation of this in the comments on the property as it's defined in the task.  This allows the version of the shared framework written to the runtimeconfig.json file to be configured independently of the package version.  Since we don't update the target framework version for patch versions of the shared framework, this is the way you would choose a patch version of the runtime.  It would also allow us to update the Microsoft.NETCore.App package so that the latest version supportss targeting all previous versions of .NET Core, by putting TargetFrameworkVersion-specific `.targets` files in the NuGet package that would set this property to the appropriate shared framework version for each supported TargetFrameworkVersion.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: This is going to be a maintenance headache. Is the plan to add conditions here for every patch release?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Should this be 1.6.1 now?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: I don't like the inconsistency. Note that there is a real use case for not referencing this package at all: standalone app with smallest possible footprint. 

Also, it's worrying that the only clean way to consume a patch release would be to acquire a new SDK.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (dasMulli) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Shouldn't this be `1.0.3` now to qualify for support? https://github.com/dotnet/core/issues/391#issuecomment-268108521

**Comment by dsplaisted** [comment]
I've updated the PR to try to better handle the differences between the targeted framework version, the version of the platform package that is referenced, and the version of the shared framework.

Currently, using version 1.1.0 of the Microsoft.NETCore.App package while targeting netstandard1.0 does not work correctly.  The project is compiled against .NET Standard 1.0, but the version of the shared framework that is written to the `runtimeconfig.json` file is 1.1.0, because we [use the version of the platform package](https://github.com/dotnet/sdk/blob/694f6bf46ad9e889627a02d5209e36ac9e2d6ee1/src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs#L86) for this.

This will generally cause F5 to fail if you update to use the later version of the Microsoft.NETCore.App package, because the app now requires version 1.1.0 of the shared framework, which isn't installed by default with the tooling.  See #493 and #501.  It can also cause issues where the wrong files are copied local because the build thinks the app will run on one shared framework but it actually runs on a different one (see dotnet/corefx#14696).

I've tried to address these issues by making the following changes:

- When targeting .NET Core, reference the version of the Microsoft.NETCore.App package that corresponds to the targeted version of .NET Core.
- Allow the version of the shared framework written to the `runtimeconfig.json` file to be overridden using the `RuntimeFrameworkVersion` MSBuild property.

## Patch Versions of the Runtime

Currently, a patch version of the runtime (for example 1.0.1 or 1.0.3) does not update the target framework version (ie it remains 1.0 for the previous examples).  If we want to automatically target a higher patch version of the runtime, we can automatically do this in the SDK.

In order to support targeting a new patch version of the runtime without an update to the SDK, the `RuntimeFrameworkVersion` can be set in the project.  This does not change the version of the shared framework package that is referenced, but it does change the version of the framework written to `runtimeconfig.json`.  This should work because I believe our policy is that patch versions do not change API surface, so the app can be compiled against the non-patched version of the runtime.

## Outstanding Questions

There are still some outstanding questions based on chats I had with @ericstj, @davidfowl, and @terrajobst.  Namely:

- Do we require a tools update to target a new version of the runtime?
- Should the latest Microsoft.NETCore.App package support targeting previous versions of the runtime, or only a single version?

I believe these changes move us in the right direction without limiting how we answer these questions.

If we want to always use the latest version of the Microsoft.NETCore.App package, then we can add `.targets` files to the package that set the `RuntimeFrameworkVersion` property to the correct value for each possible target framework.  We can also lock the tools to targeting a maximum version of the runtime (or shared framework package) if we want to explicitly disallow targeting newer runtimes than the tools.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: > standalone app with smallest possible footprint

I'm not sure if that's actually supported at all today.  If it is, then we should figure out how it fits into this, but you **can** always remove the package reference in your project:

```xml
<PackageReference Remove="Microsoft.NETCore.App" />
```

> Also, it's worrying that the only clean way to consume a patch release would be to acquire a new SDK.

You can set the `RuntimeFrameworkVersion` property to target a new patch without getting a new SDK.  See my comments: https://github.com/dotnet/sdk/pull/450#issuecomment-269834989

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: > This is going to be a maintenance headache. Is the plan to add conditions here for every patch release?

If we want to automatically target the latest patch release available, I think we have to have the mapping from TFM to patch release somewhere.  Do you see a better way to do this?

**Comment by nguerrera** [comment]
>>standalone app with smallest possible footprint

> I'm not sure if that's actually supported at all today. If it is, then we should figure out how it fits into this

It needs to be supported somehow. Microsoft.NETCore.App pulls in Roslyn + closure and things like Encoding code page data that are specifically designed to be excluded from the small standalone app category.

It is currently documented here https://docs.microsoft.com/en-us/dotnet/articles/core/preview3/deploying/ and works (I've tested manually, we are missing automated coverage).

Now, the whole 2.0 mega-package direction will probably change the mechanic to some form of automated trimming, which would make my concern moot, but we're not there yet.

>  but you can always remove the package reference in your project:
> `<PackageReference Remove="Microsoft.NETCore.App" />`

But then why the asymmetry with the other implicit refs being disable-able with the boolean switch?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: I tried this at your suggestion, and it turns out that referencing version 1.6.1 causes a bunch of extra files to be published.  I've filed #571 for this.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: @dasMulli 

> Shouldn't this be 1.0.3 now to qualify for support?

I've filed #572 for this.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (onovotny) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: FWIW, 1.6.1 is essential for Xamarin as my users have hit issues installing 1.6.0 into Xamarin Studio projects. 1.6.1 fixes this, and should IMO be the default for all libraries to prevent user pain.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (dasMulli) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Naming question: How is "default" being different from "implicit"? Currently, there is a concept of "implicit" framework references. But then there is a "default" version?
My intuition would suggest there is a version property and if not set, a default would be used. Here, only the "default" is used if I'm not mistaken.
I feel it would be clearer if the property is named `ImplicitNetCoreAppPackageVersion`. Or simply `NetCoreAppPackageVersion`. Does that make sense?

Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: @dasMulli I've renamed these properties from "Default" to "Implicit"

File: src/Templates/ProjectTemplates/CSharp/.NETCore/CSharpClassLibrary/ProjectTemplate.csproj
Comment (srivatsn) [inline src/Templates/ProjectTemplates/CSharp/.NETCore/CSharpClassLibrary/ProjectTemplate.csproj]: This ItemGroup should be deleted

File: src/Templates/ProjectTemplates/CSharp/.NETStandard/CSharpNetStandardClassLibrary/ProjectTemplate.csproj
Comment (srivatsn) [inline src/Templates/ProjectTemplates/CSharp/.NETStandard/CSharpNetStandardClassLibrary/ProjectTemplate.csproj]: This too.

**Review by srivatsn — APPROVED** [review]
Looks good modulo the extra mismatched ItemGroup in the templates.

**Comment by onovotny** [comment]
Before this gets merged, I'd strongly suggest that the default .NET Standard Library be bumped to 1.6.1. There are issues with 1.6 libraries installing into Xamarin Studio that 1.6.1 fixes.

**Comment by dsplaisted** [comment]
@onovotny Without this change the default templates will reference version 1.6.0 of NETStandard.Library.  With this change, it will be referenced by default, but you can still upgrade either in the Package Manager UI, or by adding the following to your project (which is what the UI will do if you upgrade):

```xml
<PackageReference Update="NETStandard.Library" Version="1.6.1" />
```

So this change doesn't make the situation significantly better or worse, so I'd like to handle updating the default separately since it introduces other issues (#571).

**Comment by onovotny** [comment]
@dsplaisted It doesn't make things better or worse, but it doesn't set people up in the right direction either. People won't necessarily update (or won't bother to update because they don't know there's a potential tooling issue around Xamarin that 1.6.1 fixes). That's why I'm suggesting that 1.6.1 be the version referenced by default.

I've just seen too many issues filed against my projects that pulled in 1.6.0 by Xamarin users.

**Review by nguerrera — APPROVED** [review]

**Comment by dsplaisted** [comment]
@onovotny I looked more at updating the version of NETStandard.Library to 1.6.1 and ran into some more issues.  I'm continuing to work through those but I didn't want to hold up this PR and it will probably be best to review the changes I make separately, so I've gone ahead and merged this.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props
Comment (ramarag) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props]: Apart from the project and command line, do you see that this variable getting set for specific targets.
For example The Target for cache command will not want this implicit reference

cc @schellap

**Pull request merged**
